File: Algorithm/C++/921. Minimum Add to Make Parentheses Valid.cpp
/*
 * Created on Tue Oct 23 2018 20:30:50
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// greedy, think with stack
class Solution {
public:
    int minAddToMakeValid(string S) {
      int left = 0, right = 0, result = 0;
      for(char c : S) {
        if(c == '(')
          left += 1;
        else {
          if(left == 0)
            result += 1;
          else
            left -= 1;
        }
      }
      result += left;
      return result;
    }
};

File: Algorithm/Python/921. Minimum Add to Make Parentheses Valid.py
# -*- coding: utf-8 -*-
# Created on Tue Oct 23 2018 20:28:25
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# greedy, think with stack
class Solution(object):
    def minAddToMakeValid(self, S):
        """
        :type S: str
        :rtype: int
        """
        left, right, result = 0, 0, 0
        for c in S:
            if c == '(':
                left += 1
            else:
                if left == 0:
                    result += 1
                else:
                    left -= 1
            result += left
        return result
